src/reader.rs

```diff
@@ -204,21 +146,152 @@ impl Reader {
 
         DbIterator::new(range.clone(), mem_iter, l0_iters, sr_iters).await
     }
+}
 
-    fn unwrap_value_if_not_expired(entry: &RowEntry, now_ttl: i64) -> Option<Bytes> {
-        if is_not_expired(entry, now_ttl) {
-            entry.value.as_bytes()
-        } else {
-            None
+struct LevelGet<'a> {
+    key: &'a [u8],
+    max_seq: Option<u64>,
+    snapshot: &'a (dyn ReadSnapshot + Sync + Send),
+    table_store: Arc<TableStore>,
+    db_stats: DbStats,
+    now: i64,
+    include_wal_memtables: bool,
+    include_memtables: bool,
+}
+
+impl<'a> LevelGet<'a> {
+    async fn get(&'a self) -> Result<Option<Bytes>, SlateDBError> {
+        let getters: Vec<BoxFuture<'a, Result<Option<RowEntry>, SlateDBError>>> = vec![
+            Box::pin(self.get_memtable()),
+            Box::pin(self.get_l0()),
+            Box::pin(self.get_compacted()),
+        ];
+
+        self.get_inner(getters).await
+    }
+
+    async fn get_inner(
+        &'a self,
+        getters: Vec<BoxFuture<'a, Result<Option<RowEntry>, SlateDBError>>>,
+    ) -> Result<Option<Bytes>, SlateDBError> {
+        for getter in getters {
+            let result = match getter.await? {
+                Some(result) => result,
+                None => continue,
+            };
```

Comment:
nit: I like that you broke up the memtable/l0/compacted searches into separate functions, but I feel that sticking them in a vector and iterating like this is a bit unnecessarily convoluted. Would be more clear and less susceptible to accidental reordering if we just call each function directly in #get (and then extract the `is_tombstone || is_expired` check into a helper function to handle the results of these `get_memtable/l0/compacted` calls)

Reply:
Ah I take it this was to make it easy to mock and write test cases? I guess  instead of hardcoding the three `get_xyz` calls maybe you could just pass them in as in individual parameters, eg instead of `getters: Vec<BoxFuture<'a, Result<Option<RowEntry>, SlateDBError>>>,` it would be 

```
memtable: BoxFuture<'a, Result<Option<RowEntry>, SlateDBError>>
l0: BoxFuture<'a, Result<Option<RowEntry>, SlateDBError>>
compacted: BoxFuture<'a, Result<Option<RowEntry>, SlateDBError>>
```
Just feel this improves readability somewhat and helps ensure these get called in the right ordering. But if you prefer the Vec approach I won't insist on these changes 🙂

Reply:
thank you for your feedback. yes, the intention is to make it easier to mock and to validate the logic on leveled get(). i hope to treat this leveled `get()` as general logic which accesses the levels one by one, making it testable.

if I remove the `Vec` here, I would have to place the code repeatedly, as seen in: https://github.com/slatedb/slatedb/pull/537/files/892d7e4dfc15ca55064e3657e2870f7555823826#diff-267c9f4da66412a9f439ac08d224356fe24265b5e1cebb6c44c2d55b96414513R184. to reduce duplication, I'm afraid it would still require some sort of Vec or Iterator.

by now, i think having a table driven tests is also considered as prioritized. so i hope to keep this `get_inner()` with a `getters` Vec parameter.

Reply:
sounds good